kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -112,65 +142,108 @@ else if (requestFilterResult.message() != null) {
+        if (msg instanceof InternalResponseFrame<?> decodedFrame) {
+            // jump the queue, let extra requests flow back to their sender
+            if (decodedFrame.isRecipient(filter)) {
                 if (LOGGER.isDebugEnabled()) {
                     LOGGER.debug("{}: Completing {} response for request sent by this filter{}: {}",
                             ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
                 }
-                CompletableFuture<ApiMessage> p = frame.promise();
+                CompletableFuture<ApiMessage> p = decodedFrame.promise();
                 p.complete(decodedFrame.body());
             }
             else {
-                var filterContext = new DefaultFilterContext(filter, ctx, decodedFrame, null, timeoutMs, sniHostname);
-                if (LOGGER.isDebugEnabled()) {
-                    LOGGER.debug("{}: Dispatching upstream {} response to filter {}: {}",
-                            ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
+                doRead(ctx, msg);
+            }
+        }
+        else if (readFuture.isDone()) {
+            readFuture = doRead(ctx, msg);
+        }
+        else {
+            readFuture = readFuture.whenComplete((a, b) -> {
+                if (ctx.channel().isOpen()) {
+                    doRead(ctx, msg);
                 }
-                var stage = invoker.onResponse(decodedFrame.apiKey(), decodedFrame.apiVersion(),
-                        decodedFrame.header(), decodedFrame.body(), filterContext);
-                if (stage == null) {
-                    LOGGER.error("{}: Filter{} for {} response unexpectedly returned null. This is a coding error in the filter. Closing connection.",
-                            ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
+            });
+        }
+    }
+
+    private CompletableFuture<Void> doRead(ChannelHandlerContext ctx, Object msg) {
+        if (msg instanceof DecodedResponseFrame<?> decodedFrame) {
+            var filterContext = new DefaultFilterContext(filter, ctx, decodedFrame, null, timeoutMs, sniHostname);
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("{}: Dispatching upstream {} response to filter {}: {}",
+                        ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
+            }
+            var stage = invoker.onResponse(decodedFrame.apiKey(), decodedFrame.apiVersion(),
+                    decodedFrame.header(), decodedFrame.body(), filterContext);
+            if (stage == null) {
+                LOGGER.error("{}: Filter{} for {} response unexpectedly returned null. This is a coding error in the filter. Closing connection.",
+                        ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
+                filterContext.closeConnection();
+                return CompletableFuture.completedFuture(null);
```

Comment:
I don't think so.

IMO this will be a filter author programming issue (at least, it will be once we make the javadoc clear that the the filter MUST return a CompletionStage / @NoneNull applied to the methods).   I'm actually questioning whether the error should be actually a warn.

I think we will probably want metrics to count cases where filters:

* return null 
* complete exceptionally
* filter timeouts

Obviously all work for another PR.